ci: add python e2e test workflow

ajbozarth opened the pull request:
This adds the following improvements:
- Add new `Python - E2E Tests` workflow that will run when a push to `main` includes changes in `/python/`
    - similar to the `TypeScript - E2E Examples` workflow, it installs Ollama to run tests
    - this can be extended later with env vars to also test examples using other model services like watsonx
- Document that `poetry install --all-extras` is usually needed for contributing developers
- Add `poe docs --type check` to pre-commit
- Add missing unit test decorators to `tests/backend/providers/test_azure_openai.py`

AngeloDanducci commented on python/CONTRIBUTING.md:
```diff
@@ -85,7 +85,7 @@ cd beeai-framework
 
 # Use Poetry to install the project dependencies and activate a virtual environment
-poetry install
+poetry install # poetry install --all-extras
```
Given the mypy problems we saw in #576 should all extras be on a separate line so it appears mandatory?

The way it currently is makes it appear optional to me.

ajbozarth replied:
I did it this was since that was how we did `poe format --fix above` and not everyone following these step plan to contribute code, some just want a local build to play with the cutting edge. But I'm ok with just replacing `poetry install` with `poetry install --all-extras entirely here

Tomas2D reviewed (commented):
- Running examples (testing the whole flow) are E2E tests and not integration tests. 
- Running all of them for each PR is unsustainable (free tier).

If we want to really test some E2E tests (examples) then I would pick just a small subset of them.

ajbozarth commented:
So so based on review I have streamlined this PR and updated the top level description to match.

I've reverted all the e2e vs integration changes and remove e2e tests from the normal python workflow. 

Now e2e tests as a whole, including ones that need Ollama, are run in a separate workflow that only runs on `main`. This will enable discovering e2e tests failing after a push to `main` without running them all on every push to a PR.

As this is a less intrusive update we can probably merge as is
